# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationParent.cpp
# Commit: 10bb5f41f44b
# Full Hash: 10bb5f41f44b6502e1eea380f5d0e02b4da4f3d8
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2024-11-12 09:32:50
# Regressor Bug: 1928702
# File Overlap Count: 1
# Description:
#   Bug 1928702 - Part 8: Implement permission check r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D227693
# ==============================================================================

diff -r 0f64abe35d62 -r 10bb5f41f44b dom/notification/NotificationParent.cpp
--- a/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:13 2024 +0000
+++ b/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:14 2024 +0000
@@ -61,6 +61,20 @@
 mozilla::ipc::IPCResult NotificationParent::RecvShow(ShowResolver&& aResolver) {
   mResolver.emplace(std::move(aResolver));
 
+  // Step 4.1: If the result of getting the notifications permission state is
+  // not "granted", then queue a task to fire an event named error on this, and
+  // abort these steps.
+  NotificationPermission permission = GetNotificationPermission(
+      mPrincipal, mEffectiveStoragePrincipal, mIsSecureContext,
+      PermissionCheckPurpose::NotificationShow);
+  if (permission != NotificationPermission::Granted) {
+    CopyableErrorResult rv;
+    rv.ThrowTypeError("Permission to show Notification denied.");
+    mResolver.take().value()(rv);
+    mDangling = true;
+    return IPC_OK();
+  }
+
   // Step 4.2: Run the fetch steps for notification. (Will happen in
   // nsIAlertNotification::LoadImage)
   // Step 4.3: Run the show steps for notification.
@@ -154,6 +168,11 @@
 }
 
 void NotificationParent::ActorDestroy(ActorDestroyReason aWhy) {
+  if (mDangling) {
+    // We had no permission, so nothing to clean up.
+    return;
+  }
+
   nsAutoString alertName;
   GetAlertName(alertName);
   UnregisterNotification(mPrincipal, mId, alertName, CloseMode::InactiveGlobal);